src/com/amazon/ion/impl/IonCursorBinary.java

```diff
@@ -0,0 +1,721 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.amazon.ion.impl;
+
+import com.amazon.ion.BufferConfiguration;
+import com.amazon.ion.IonBufferConfiguration;
+import com.amazon.ion.IonException;
+import com.amazon.ion.IonCursor;
+import com.amazon.ion.IonType;
+import com.amazon.ion.IvmNotificationConsumer;
+
+import java.nio.ByteBuffer;
+
+/**
+ * An IonCursor over binary Ion data, capable of buffering or skipping Ion values at any depth. Records byte
+ * indices of the value currently buffered, enabling parsers direct access to the bytes in the value's representation.
+ */
+class IonCursorBinary implements IonCursor {
+
+    private static final int LOWER_SEVEN_BITS_BITMASK = 0x7F;
+    private static final int VALUE_BITS_PER_VARUINT_BYTE = 7;
+    private static final int IVM_START_BYTE = 0xE0;
+    private static final int IVM_FINAL_BYTE = 0xEA;
+    private static final int SINGLE_BYTE_MASK = 0xFF;
+    private static final int LIST_TYPE_ORDINAL = IonType.LIST.ordinal();
+    private static final IvmNotificationConsumer NO_OP_IVM_NOTIFICATION_CONSUMER = (x, y) -> {};
+
+    // Initial capacity of the stack used to hold ContainerInfo. Each additional level of nesting in the data requires
+    // a new ContainerInfo. Depths greater than 8 will be rare.
+    private static final int CONTAINER_STACK_INITIAL_CAPACITY = 8;
+
+    /**
+     * Stack to hold container info. Stepping into a container results in a push; stepping out results in a pop.
+     */
+    protected Marker[] containerStack = new Marker[CONTAINER_STACK_INITIAL_CAPACITY];
+
+    /**
+     * The index of the current container in `containerStack`.
+     */
+    protected int containerIndex = -1;
+
+    /**
+     * The Marker representing the parent container of the current value.
+     */
+    protected Marker parent = null;
+
+    /**
+     * The start offset into the user-provided byte array, or 0 if the user provided an InputStream.
+     */
+    private final long startOffset;
+
+    /**
+     * The index of the next byte in the buffer that is available to be read. Always less than or equal to `limit`.
+     */
+    private long offset;
+
+    /**
+     * The index at which the next byte received will be written. Always greater than or equal to `offset`.
+     */
+    private long limit;
```

Comment:
Some of these are physical indexes into the byte array, and others (like the Marker's endIndex) are logical indexes that may extend beyond the bounds of the physical buffer. Rather than use `int` for the former and `long` for the latter, I ended up just using `long` everywhere. I do recall experimenting with using `int` for physical indexes and finding that had no impact on the performance characteristics. Either way we do it, we have to do narrowing casts in certain situations.